### PR TITLE
fix(DataStore): [DRAFT] cascade delete parent/child

### PR DESCRIPTION
**Pull request by wooj2**
**tldr; -- Don't review this PR**.  This PR is superseded by:
https://github.com/aws-amplify/amplify-ios/pull/986
https://github.com/aws-amplify/amplify-ios/pull/987

~This PR started out as demonstrating synchronous unit tests to demonstrate improved readability.  Notice:~
* ~Absence of parent base class~
* ~XCTAsserts ONLY belong in the actual test~
* ~Ability to linearly read from the top of the function to the bottom~

~However, this PR has grown into the first version of cascade delete, to address [issue #993](https://github.com/aws-amplify/amplify-ios/issues/933). This PR will be here for historical purposes, but will work on breaking this PR up into separate PRs in order to make it easier to review.  I will break this up into two PRs:
**1st PR:** Refactor - move `queryAndDeleteTransaction` and `collpaseResults`, `syncDeletion`, into an extension (with no functional changes) 
**2nd PR:**
a.  Addition of `recurseQueryAssociatedModels` w/ unit tests
b.  Removing `private` from  `submitToSyncEngine` -- i prefer not to do this, but StorageEngine is getting large, and syncDeletion methods need to be able to call this function.~

~Main reason for doing splitting up this PR is to clearly see the diff of the implementation of `queryAndDeleteTransaction` and `collapseResults` in a single file.~

~Note that this PR relies on: https://github.com/aws-amplify/amplify-ios/pull/965~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Storage/StorageEngineTests.swift [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        guard case .success(let comments) =
+            queryModelSynchronous(modelType: Comment.self, predicate: Comment.keys.post == savedPost1.id) else {
+            XCTFail("Failed to query comment")
+            return
+        }
+        XCTAssertEqual(comments.count, 2)
+    }
+
+    /**
+     * Below are synchronous conveinence methods.  Please do not add any calls to XCTFail()
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Storage/StorageEngineTests.swift:89]: removing all calls to XCTFail is good addition! it makes it easier to move out of the test class

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Storage/StorageEngineTests.swift [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+     * in these conveinence methods.  Failures should be handled in the body of the unit test.
+     */
+    func saveModelSynchronous<M: Model>(model: M) -> DataStoreResult<M> {
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Storage/StorageEngineTests.swift:92]: i wonder how well this translate to API results as that results a success/failure and inside the success is a graphQL response with the data or graphql error that has transformation/partial/error cases. This nested response means that the helper for API requests will result in aggregating the results in the helper method and returning `Result<M>` so that the caller can simply do `guard case .success(let m) = ..`

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    func resolveAssociatedModels<M: Model>(_ result: DataStoreResult<([M], [Model])>) -> [Model] {
+        switch result {
+        case .success(let results):
+            return results.1
+        case .failure:
+            return []
+        }
+    }
+
+    func generateMutationsFromModels(models: [Model]) -> [MutationEvent] {
```
Comment (wooj2) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine+DeleteTransaction.swift:142]: Remove, no longer needed.
